src/RazorSdk/SourceGenerators/IncrementalRazorSourceGenerator.cs

```diff
@@ -0,0 +1,123 @@
+
+            var references = context.CompilationProvider
+                .WithLambdaComparer(
+                    (c1, c2) => c1 != null && c2 != null && c1.References != c2.References)
+                .Select((compilation, ct) => compilation.References);
+
+            var discoveryProjectEngine = references
+                .Combine(rsgOptions)
+                .Combine(sourceItems.Collect())
+                .Select((pair, ct) =>
+                {
+                    var ((references, rsgOptions), projectItems) = pair;
+                    var tagHelperFeature = new StaticCompilationTagHelperFeature();
+                    return GetDiscoveryProjectEngine(tagHelperFeature, references, projectItems, rsgOptions.RootNamespace);
+                });
+
+            var syntaxTrees = sourceItems
+                .Combine(discoveryProjectEngine)
+                .Select((pair, ct) =>
+                {
+                    var (item, discoveryProjectEngine) = pair;
+
+                    var codeGen = discoveryProjectEngine.Process(item);
+                    var generatedCode = codeGen.GetCSharpDocument().GeneratedCode;
+                    return CSharpSyntaxTree.ParseText(generatedCode, new CSharpParseOptions(LanguageVersion.Preview));
+                });
+
+            var tagHelpersFromCompilation = syntaxTrees
+                .Combine(context.CompilationProvider)
+                .Combine(discoveryProjectEngine)
+                .Select((pair, ct) =>
+                {
+                    var ((syntaxTrees, compilation), discoveryProjectEngine) = pair;
+                    var tagHelperFeature = GetFeature<StaticCompilationTagHelperFeature>(discoveryProjectEngine);
+                    return GetTagHelpersFromCompilation(
+                        compilation,
+                        tagHelperFeature ?? new StaticCompilationTagHelperFeature(),
+                        syntaxTrees
+                    );
+                });
+
+            var tagHelpersFromReferences = discoveryProjectEngine
+                .Combine(context.CompilationProvider)
+                .WithLambdaComparer<(RazorProjectEngine, Compilation)>(
+                    (p1, p2) => p1.Item2.References != p2.Item2.References
+                )
+                .Select((pair, ct) =>
+                {
+                    var (discoveryProjectEngine, compilation) = pair;
+                    var tagHelperFeature = GetFeature<StaticCompilationTagHelperFeature>(discoveryProjectEngine);
+                    return GetTagHelpers(
+                        compilation.References,
+                        tagHelperFeature ?? new StaticCompilationTagHelperFeature(),
+                        compilation
+                    );
+                });
+
+            var tagHelpers = tagHelpersFromCompilation.Combine(tagHelpersFromReferences)
+                .WithLambdaComparer((tagHelpers, newTagHelpers) =>
```

Comment:
Yeah, this isn't actually that helpful. We'll always end up generating a new update to the tag helpers provider since the AdditionalTexts are going to be modified since we're doing the compilation.